WalletWasabi/Services/WalletService.cs

```diff
@@ -802,6 +732,97 @@ public async Task<Block> GetOrDownloadBlockAsync(uint256 hash, CancellationToken
+						}
+
+						var localEndPoint = ServiceConfiguration.BitcoinCoreEndPoint;
+						var localNode = await Node.ConnectAsync(Network, localEndPoint, nodeConnectionParameters);
+						try
+						{
+							Logger.LogInfo<WalletService>("TCP Connection succeeded, handshaking...");
+							localNode.VersionHandshake(Constants.LocalNodeRequirements, handshakeTimeout.Token);
+							var peerServices = localNode.PeerVersion.Services;
+
+							//if (!peerServices.HasFlag(NodeServices.Network) && !peerServices.HasFlag(NodeServices.NODE_NETWORK_LIMITED))
+							//{
+							//	throw new InvalidOperationException($"Wasabi cannot use the local node because it does not provide blocks.");
+							//}
+
+							Logger.LogInfo<WalletService>($"Handshake completed successfully.");
+
+							if (!localNode.IsConnected)
+							{
+								throw new InvalidOperationException($"Wasabi could not complete the handshake with the local node and dropped the connection.{Environment.NewLine}" +
+									$"Probably this is because the node does not support retrieving full blocks or segwit serialization.");
+							}
+							LocalBitcoinCoreNode = localNode;
+						}
+						catch (OperationCanceledException) when (handshakeTimeout.IsCancellationRequested)
+						{
+							Logger.LogWarning<Node>($"Wasabi could not complete the handshake with the local node. Probably Wasabi is not whitelisted by the node.{Environment.NewLine}" +
+								$"Use \"whitebind\" in the node configuration. (Typically whitebind=127.0.0.1:8333 if Wasabi and the node are on the same machine and whitelist=1.2.3.4 if they are not.)");
+							throw;
+						}
+					}
+				}
+
+				// Get Block from local node
+				Block blockFromLocalNode = null;
+				// Should timeout faster. Not sure if it should ever fail though. Maybe let's keep like this later for remote node connection.
+				using (var cts = new CancellationTokenSource(TimeSpan.FromSeconds(64)))
+				{
+					blockFromLocalNode = await LocalBitcoinCoreNode.DownloadBlockAsync(hash, cts.Token);
+				}
+
+				// Validate retrieved block
+				if (!blockFromLocalNode.Check())
+				{
+					throw new InvalidOperationException("Disconnected node, because invalid block received!");
+				}
+
+				// Retrieved block from local node and block is valid
+				Logger.LogInfo<WalletService>($"Block acquired from local P2P connection: {hash}");
+				return blockFromLocalNode;
+			}
+			catch (Exception ex)
+			{
+				DisconnectDisposeNullLocalBitcoinCoreNode();
+
+				if (ex is SocketException)
+				{
+					Logger.LogTrace<WalletService>("Did not find local listening and running full node instance. Trying to fetch needed block from other source.");
+				}
+				else
+				{
+					Logger.LogWarning<WalletService>(ex);
+				}
+			}
+			cancel.ThrowIfCancellationRequested();
+
+			return null;
+		}
```

Comment:
This whole part is just copypaste, right?

Reply:
Yes